pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "egresspolicy")
+)
+
+// The egresspolicy manager stores the internal data tracking the policy
+// and pod mappings. And also hooks up all the callbacks to update the egress
+// map accordingly.
+type Manager struct {
+	mutex lock.Mutex
+
+	// Stores pod to policy mapping
+	policyPods map[podID][]policyID
+	// Stores policy configs indexed by policyID
+	policyConfigs map[policyID]*EgressPolicyConfig
+	// Stores podId to pod metadata mapping
+	podDataStore map[podID]*podMetadata
+
+	storeGetter StoreGetter
+}
+
+func NewEgressPolicyManager() *Manager {
+	return &Manager{
+		policyPods:    make(map[podID][]policyID),
+		policyConfigs: make(map[policyID]*EgressPolicyConfig),
+		podDataStore:  make(map[podID]*podMetadata),
+	}
+}
+
+type StoreGetter interface {
+	GetStore(name string) cache.Store
+}
+
+func (manager *Manager) RegisterGetStores(sg StoreGetter) {
+	manager.storeGetter = sg
+}
+
+// Event handlers
+
+// AddEgressPolicy parses the given policy config, and updates internal state with the config fields.
+// returns bool indicates if policy is added, err inidates first encountered error
+func (manager *Manager) AddEgressPolicy(config EgressPolicyConfig) (bool, error) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	_, ok := manager.policyConfigs[config.id]
+	if ok {
+		log.WithField(
+			logfields.CiliumEgressNATPolicyName, config.id.Name).Warn("CiliumEgressNATPolicy already exists and is not re-added")
+		return false, errors.New("already exists")
+	}
+
+	err := manager.isValidConfig(config)
+	if err != nil {
+		return false, fmt.Errorf("invalid config: %w", err)
+	}
+
+	manager.policyConfigs[config.id] = &config
+	pods := manager.getLocalPodsForPolicy(&config)
+	err = manager.upsertPolicyPod(&config, pods...)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Returns a slice of endpoint pods metadata that are selected by the given policy config.
+func (manager *Manager) getLocalPodsForPolicy(config *EgressPolicyConfig) []*podMetadata {
+	var (
+		retPods []*podMetadata
+		podData *podMetadata
+		err     error
+	)
+
+	podStore := manager.storeGetter.GetStore("pod")
+	for _, podItem := range podStore.List() {
+		pod, ok := podItem.(*slim_corev1.Pod)
+		if !ok {
+			continue
+		}
+		podIPs := k8sUtils.ValidIPs(pod.Status)
+		if len(podIPs) == 0 {
+			continue
+		}
+		if podData, err = manager.getPodMetadata(pod, podIPs); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.K8sPodName:   pod.Name,
+				logfields.K8sNamespace: pod.Namespace,
+			}).Error("Failed to get valid pod metadata")
+			continue
+		}
+		if k8sUtils.GetLatestPodReadiness(pod.Status) != slim_corev1.ConditionTrue {
+			continue
+		}
+		if !config.policyConfigSelectsPod(podData) {
+			continue
+		}
+		retPods = append(retPods, podData)
```

Comment:
Why are we duplicating information in `podData` if we could simply return a slice of pod pointers, which has this information, that are selected by this policy?

Reply:
This point probably don't apply anymore since we are watching the cilium endpoints now and we rely on the cache we build internally now.